app/services/foreman/provision/ssh.rb

```diff
@@ -1,45 +1,73 @@
-require 'timeout'
+require 'open3'
 
 class Foreman::Provision::Ssh
-  attr_reader :template, :uuid, :results, :address, :username, :options
-
-  def initialize(address, username = "root", options = { })
+  attr_reader :template, :uuid, :address, :username, :options
+
+  # options: {
+  #   template: "/path/to/template",
+  #   uuid:     "uuid",
+  #   key_data: ["private-key-content"],
+  #   password: "string",
+  # }
+  def initialize(address, username = "root", options = {})
     @username = username
     @address  = address
     @template = options.delete(:template) || raise("must provide a template")
     @uuid     = options.delete(:uuid)     || "#{address}-#{username}"
-    @options  = defaults.merge(options)
+    @options  = options
 
     initiate_connection!
   end
 
   def deploy!
-    logger.debug "about to upload #{template} to remote system at #{remote_script}"
-    scp.upload(template, remote_script)
-    logger.debug "about to execute #{command}"
-    @results = ssh.run(command)
-    log_stdout
-    log_stderr
-    success?
+    scp(template)
+
+    logger.debug "about to execute #{main_command}"
+
+    stdout, stderr, status = Open3.capture3(ssh_cmd(main_command))
+
+    logger.debug(stdout) if stdout.presence
+    logger.error(stderr) if stderr.presence
+
+    status.success?
   end
 
   private
 
-  def success?
-    return true if results.empty?
-    results.map(&:status).compact == [0]
+  def ssh_auth_method
+    if @options[:key_data].present? && @options[:key_data].any?
+      :key
+    elsif @options[:password].present?
+      :password
+    else
```

Comment:
Wouldn't it be better to set priority in PreferredAuthentications based on what we have and let ssh deal with it?